Draft `getFederatedType` for Cost-Based Federated Planner

min-guk opened the pull request:
This PR introduces a draft of the `getFederatedType` function in `FederatedPlanRewireTransTable.java`, which determines and propagates `FType` in the cost-based federated planner.

The function integrates and extends `allowFederated` and `getFederatedOut` from `AFederatedPlanner`, aiming to support more operations.

Referenced:
- `sysds/hops/*Op.java`
- `sysds/runtime/instructions/fed/*FEDInstruction.java`
- `sysds/hops/fedplanner/AFederatedPlanner.java`
- `sysds/hops/fedplanner/FederatedPlannerFedAll.java`

Currently, most `FType` values returned by `getFederatedType` are still `null`, so the implementation is likely insufficient and incomplete. Please review whether the logic correctly handles opcodes and input types.

**Note:**
- This PR should ideally be split, but wasn't due to time constraints.
- Many tests are currently failing.
- I plan to cancel this PR and re-submit smaller ones after review.

Thanks in advance for your feedback!

j143 commented on src/main/java/org/apache/sysds/hops/fedplanner/FederatedMemoTable.java:
```diff
@@ -100,54 +136,88 @@ public static class FedPlanVariants {
 		protected List<FedPlan> _fedPlanVariants;  // List of plan variants
 
+		public FedPlanVariants(HopCommon hopCommon, FederatedOutput fedOutType) {
+			this.hopCommon = hopCommon;
```
Hi, here the syntax seems to be incorrect.

j143 commented on src/main/java/org/apache/sysds/hops/fedplanner/FederatedMemoTable.java:
```diff
@@ -28,16 +28,21 @@
 	private final Map<Pair<Long, FederatedOutput>, FedPlanVariants> hopMemoTable = new HashMap<>();
 
+	public void addFedPlanVariants(long hopID, FederatedOutput fedOutType, FedPlanVariants fedPlanVariants) {
+		hopMemoTable.put(new ImmutablePair<>(hopID, fedOutType), fedPlanVariants);
```
Hi, here this seems to be syntax issue - most likely due to rebase, I guess.

min-guk commented:
@j143  The compile and build issues should be resolved now.

min-guk commented:
@mboehm7 One question: during FType propagation, if a TransRead has multiple TransWrite children and their FTypes conflict, how should that be handled?

Is there any rule or priority for resolving such conflicts?

```
// FederatedKMeansPlanningTest
[Error] FType MISMATCH DETECTED for TransRead (hopName: C, hopID: 775)
[Error] TRANSREAD HOP DETAILS - Type: DataOp, OpType: TRead, DataType: MATRIX, Dims: [4x-1]
[Error] FILTERED CHILD HOPS FTYPE ANALYSIS:
[Error]   FilteredChild[0] - Name: C, ID: 644, FType: ROW, Type: DataOp, OpType: TWrite, Dims: [4x100]
[Error]   FilteredChild[1] - Name: C, ID: 696, FType: null, Type: DataOp, OpType: TWrite, Dims: [4x1000]
[Error] Expected FType: ROW, Mismatched FType: null at child index: 1
```

min-guk commented:
I removed unnecessary modifications to clean up the PR, fixed the previous compilation issues, and deleted incomplete test code.

min-guk commented:
I’ve made the following changes. Apologies for the repeated updates, and thank you again for your help in testing.

> LicenseCheck

Added missing license headers to newly added test scripts.

> Documentation / Java
Python Test

Fixed incorrect annotations for method input parameters in FederatedPlanCostEnumerator.

> JavaTest/**.component.e**.**,**.component.f**.**,**.component.m**.**

Modified FederatedPlanCostEnumeratorTest to avoid invoking the PythonVisualizer.

> JavaTest/ **.functions.a**.**,**.functions.binary.matrix.**,**.functions.binary.scalar.**,**.functions.binary.tensor.**
> JavaTest/**.functions.federated.primitives.part2.** -Dtest-threadCount=1 -Dtest-forkCount=1
```
unable to write test matrix (...) : Destination exists and is not a directory
```
This was likely caused by residual files from a previous FederatedPlanCostEnumeratorTest run where output files were not properly deleted. I've updated the test to avoid redirecting terminal output to a file, which should prevent the conflict.

min-guk commented:
Hi @mboehm7, 
I’ve added the missing test script that I had overlooked earlier. Apologies for repeatedly asking you to rerun the tests.

Also, I had a quick question: from what I can tell, the remaining test failures don’t seem to be caused by my recent changes — but could you confirm if that’s indeed the case?

> JavaTest/**.component.e**.**,**.component.f**.**,**.component.m**.**

- Added the previously missing test script (FederatedPlanCostEnumeratorTest12.dml).

> JavaTest/ **.functions.a**.**,**.functions.binary.matrix.**,**.functions.binary.scalar.**,**.functions.binary.tensor.**

- ElementwiseAdditionMultiplicationTest is failing due to a file/directory conflict.

> JavaTest/**.functions.federated.primitives.part2.** -Dtest-threadCount=1 -Dtest-forkCount=1

- FederatedLeftIndexTest is failing when attempting to connect workers on dynamic ports (e.g., 34595, 41477) on localhost.

min-guk commented:
Hi @mboehm7, I’ve resolved the error that was caused by a typo in `FederatedPlanCostEnumeratorTest12.dml`.

Could you please take a look at the `getFederatedType` function at line 619 in
`src/main/java/org/apache/sysds/hops/fedplanner/FederatedPlanRewireTransTable.java`?

In the meantime, I’ll also review and debug the function on my end.
